### PR TITLE
Port CA1806 ("Do not ignore method results") as an IOperation analyzer.

### PR DESCRIPTION
**Pull request by mavasani**

### CODE REVIEW

**Comment by mavasani** [comment]
@JohnHamby @srivatsn @genlu @nguerrera @lgolding

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs [hunk 1]
```diff
@@ -1,68 +1,146 @@
...
+                            {
+                                messageFormat = MicrosoftMaintainabilityAnalyzersResources.DoNotIgnoreMethodResultsMessageObjectCreation;
+                                targetMethodName = ctor.ContainingType.Name;
+                            }
+                            break;
+
+                        case OperationKind.InvocationExpression:
+                            var invocationExpression = ((IInvocationExpression)expression);
+                            var targetMethod = invocationExpression.TargetMethod;
+                            if (targetMethod == null)
```
Comment (genlu) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: In which case the target method will be null?

Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: I wanted to actually ask @JohnHamby if it is possible. If it is not possible, I can remove their null check.

**Comment by srivatsn** [comment]
LGTM

**Comment by genlu** [comment]
:+1:

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs [hunk 2]
```diff
@@ -1,68 +1,146 @@
...
+                "Join",
+                "Normalize",
+                "Remove",
+                "Replace",
+                "Split",
+                "PadLeft",
+                "PadRight",
+                "Substring",
+            });
+
```
Comment (nguerrera) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: Do we have an issue tracking extending this to anything with [PureAttribute] applied?

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs [hunk 3]
```diff
@@ -1,68 +1,146 @@
...
+                s_stringMethodNames.Contains(method.Name);
+        }
+
+        private static bool IsTryParseMethod(IMethodSymbol method)
+        {
+            return method.Name.StartsWith("TryParse", StringComparison.OrdinalIgnoreCase) &&
+                method.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                method.Parameters.Length >= 2 &&
+                method.Parameters[1].RefKind != RefKind.None;
+        }
```
Comment (nguerrera) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: [Aside: I have been occasionally guilty of relying on known TryParse implementation returning default(T) and ignoring the bool in limited scenarios where "or-default" is what I want anyhow, but this is good practice to enforce.]

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: I don't think it is possible for semantically well-formed code. I don't know if there are cases where one of the compilers can generate a bound invocation with a null target in the presence of errors.

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs [hunk 4]
```diff
@@ -1,68 +1,146 @@
...
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182273.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(ObjectCreationRule, StringCreationRule, HResultOrErrorCodeRule, TryParseRule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterOperationBlockStartAction(osContext =>
```
Comment (JohnHamby) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: Does the "os" in "osContext" mean anything?

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs (hunk 1 above)
Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: Thanks John, I'll retain the check for now.

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs (hunk 2 above)
Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: Added a note: https://github.com/dotnet/roslyn-analyzers/issues/746#issuecomment-163084601

File: src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs (hunk 4 above)
Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs]: No, just a shorthand for operation start block context.

**Pull request merged**